### ISSUE TITLE
Crash when calling `invocations.append` from multiple threads

### ISSUE THREAD

**Issue by zmeziane** [MakeAWishFoundation/SwiftyMocky#272]
### Problem
Crash when calling `invocations.append` from multiple threads:

![image](https://user-images.githubusercontent.com/3857761/103096615-cabb0600-4604-11eb-802c-06d8b8666dd9.png)
```
xctest(33348,0x70000e07b000) malloc: pointer 0x7fa8eaa30000 being freed was not allocated
xctest(33348,0x70000e07b000) malloc: *** set a breakpoint in malloc_error_break to debug
xctest(33348,0x70000e07b000) malloc: pointer 0x7fa8eaa30000 being freed was not allocated
```

### Expected behaviour

Should not crash 😄 . Probably need to make reading/writing to `invocations` thread safe.

### Environment
- SwiftyMocky - 3.5.0
- Xcode - 12.3
- macOS - 10.15.7
- 
#### Input
```swift
import XCTest
import SwiftyMocky

//sourcery: AutoMockable
protocol Fetcher {
    func fetchProperty(with id: String) -> String
}

struct CombinedFetcher {
    let fetcher: Fetcher

    func fetch(forIds ids: [String], _ completion: @escaping () -> Void) {
        let queue = DispatchQueue(label: "test", attributes: .concurrent)
        let group = DispatchGroup()

        ids.forEach { id in
            group.enter()
            queue.async {
                _ = fetcher.fetchProperty(with: id)
                group.leave()
            }
        }

        group.notify(queue: DispatchQueue.main) {
            completion()
        }
    }
}

class CombinedFetcherTests: XCTestCase {
    
    func test_fetch() {
        let id1 = "id1"
        let id2 = "id2"
        let fetcher = FetcherMock()
        Given(fetcher, .fetchProperty(with: .value(id1), willReturn: ""))
        Given(fetcher, .fetchProperty(with: .value(id2), willReturn: ""))
        let sut = CombinedFetcher(fetcher: fetcher)

        (0...100000).forEach { _ in
            let exp = expectation(description: "Fetch complete")
            sut.fetch(forIds: [id1, id2]) { exp.fulfill() }
            waitForExpectations(timeout: 2)
        }
    }
}

```

**Comment by g-amichnia** [comment]
Thanks for the example, I will check this. Definitely something needs to be done here.

**Comment by g-amichnia** [comment]
@zmeziane fixed in 4.0.4

**Comment by zmeziane** [comment]
@g-amichnia awesome! Thank you 🙏
